torchvision/prototype/datasets/_builtin/sbd.py

```diff
@@ -0,0 +1,179 @@
+        boundaries = (
+            torch.as_tensor(np.stack([raw_boundary[0].toarray() for raw_boundary in raw_boundaries]))
+            if decode_boundaries
+            else None
+        )
+        segmentation = torch.as_tensor(raw_segmentation) if decode_segmentation else None
+
+        return boundaries, segmentation
+
+    def _collate_and_decode_sample(
+        self,
+        data: Tuple[Tuple[Any, Tuple[str, io.IOBase]], Tuple[str, io.IOBase]],
+        *,
+        config: DatasetConfig,
+        decoder: Optional[Callable[[io.IOBase], torch.Tensor]],
+    ) -> Dict[str, Any]:
+        split_and_image_data, ann_data = data
+        _, image_data = split_and_image_data
+        image_path, image_buffer = image_data
+        ann_path, ann_buffer = ann_data
+
+        image = decoder(image_buffer) if decoder else image_buffer
+
+        if config.boundaries or config.segmentation:
+            boundaries, segmentation = self._decode_ann(
+                read_mat(ann_buffer), decode_boundaries=config.boundaries, decode_segmentation=config.segmentation
+            )
+        else:
+            boundaries = segmentation = None
+
+        return dict(
+            image_path=image_path,
+            image=image,
+            ann_path=ann_path,
+            boundaries=boundaries,
+            segmentation=segmentation,
+        )
+
+    def _make_datapipe(
+        self,
+        resource_dps: List[IterDataPipe],
+        *,
+        config: DatasetConfig,
+        decoder: Optional[Callable[[io.IOBase], torch.Tensor]],
+    ) -> IterDataPipe[Dict[str, Any]]:
+        archive_dp, extra_split_dp = resource_dps
+
+        archive_dp = resource_dps[0]
+        archive_dp = TarArchiveReader(archive_dp)
+        split_dp, images_dp, anns_dp = Demultiplexer(
+            archive_dp,
+            3,
+            self._classify_archive,  # type: ignore[arg-type]
+            buffer_size=INFINITE_BUFFER_SIZE,
+            drop_none=True,
+        )
+
+        if config.split == "train_noval":
+            split_dp = extra_split_dp
+        # TODO: replace the .map() call with decode=True in LineReader when
+        #  https://github.com/pytorch/data/issues/28 is resolved
+        split_dp = LineReader(split_dp).map(bytes.decode, input_col=1)
+        split_dp = Shuffler(split_dp)
+
+        dp = split_dp
+        for level, data_dp in enumerate((images_dp, anns_dp)):
+            dp = KeyZipper(
+                dp,
+                data_dp,
+                key_fn=getitem(*[0] * level, 1),
+                ref_key_fn=path_accessor("stem"),
+                buffer_size=INFINITE_BUFFER_SIZE,
+            )
+        return Mapper(dp, self._collate_and_decode_sample, fn_kwargs=dict(config=config, decoder=decoder))
+
+    def generate_categories_file(self, root: Union[str, pathlib.Path]) -> None:
+        dp = self.resources(self.default_config)[0].to_datapipe(pathlib.Path(root) / self.name)
```

Comment:
As resources are just fixed files, just wondering why this function is attached to Dataset class. Do we expect users would call this method in the future to generate the category files?

Reply:
Note that user will never see a `Dataset` object during "normal" operation. The public API `torchvision.prototype.datasets.load` returns a datapipe

https://github.com/pytorch/vision/blob/b1ef19639ee845902ecf60930757d028f1e1491c/torchvision/prototype/datasets/_api.py#L74